assemblyscript/assemblyscript.go

```diff
@@ -0,0 +1,163 @@
+package assemblyscript
+
+import (
+	"context"
+	"crypto/rand"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"math"
+	"os"
+	"time"
+	"unicode/utf16"
+
+	"github.com/tetratelabs/wazero"
+	"github.com/tetratelabs/wazero/api"
+)
+
+// Instantiate instantiates a module implementing abort, trace, and seed for use from AssemblyScript programs.
+// The instantiated module will output abort messages to os.Stderr, not output trace messages, and use
+// crypto/rand as the source for seed values. If the AssemblyScript program is configured to use WASI, by
+// calling "import wasi" in any file, these functions will not be used.
+func Instantiate(ctx context.Context, r wazero.Runtime) (api.Closer, error) {
+	return NewModuleBuilder().Instantiate(ctx, r)
+}
+
+// ModuleBuilder allows configuring the module that will export functions used automatically by AssemblyScript.
+type ModuleBuilder interface {
+	// WithAbortWriter sets the io.Writer that should be used to write abort messages to.
+	// Abort messages are written when an error is thrown within an AssemblyScript program.
+	//
+	// Defaults to os.Stderr.
+	WithAbortWriter(writer io.Writer) ModuleBuilder
+
+	// WithTraceWriter sets the io.Writer that should be used to write trace messages to.
+	// Trace messages are written when the built-in trace function is called within an
+	// AssemblyScript program.
+	//
+	// Defaults to io.Discard meaning trace messages are not written.
+	WithTraceWriter(writer io.Writer) ModuleBuilder
+
+	// WithSeedSource sets the io.Reader to read bytes from for seeding random number generation.
+	//
+	// Defaults to crypto/rand.Reader to seed using cryptographically random bytes.
+	WithSeedSource(reader io.Reader) ModuleBuilder
+
+	// Instantiate instantiates the module so that AssemblyScript can import from it.
+	Instantiate(ctx context.Context, runtime wazero.Runtime) (api.Closer, error)
+}
+
+// NewModuleBuilder returns a ModuleBuilder for configuring a AssemblyScript host module.
+func NewModuleBuilder() ModuleBuilder {
+	return &moduleBuilder{
+		abortWriter: os.Stderr,
+		traceWriter: io.Discard,
+		seedSource:  rand.Reader,
+	}
+}
+
+type moduleBuilder struct {
+	abortWriter io.Writer
+	traceWriter io.Writer
+	seedSource  io.Reader
+}
+
+// WithAbortWriter implements ModuleBuilder.WithAbortWriter
+func (m *moduleBuilder) WithAbortWriter(writer io.Writer) ModuleBuilder {
+	m.abortWriter = writer
+	return m
+}
+
+// WithTraceWriter implements ModuleBuilder.WithTraceWriter
+func (m *moduleBuilder) WithTraceWriter(writer io.Writer) ModuleBuilder {
+	m.traceWriter = writer
+	return m
+}
+
+// WithSeedSource implements ModuleBuilder.WithSeedSource
+func (m *moduleBuilder) WithSeedSource(reader io.Reader) ModuleBuilder {
+	m.seedSource = reader
+	return m
+}
+
+// Instantiate implements ModuleBuilder.Instantiate
+func (m *moduleBuilder) Instantiate(ctx context.Context, runtime wazero.Runtime) (api.Closer, error) {
+	return runtime.NewModuleBuilder("env").
```

Comment:
`env` is not only used by AS but is kind of "default" host functions, so I'm not sure users want to use "env" module only for assembly script or not.. maybe not defining module builder, but instead just move all of the code in this PR under an example? @codefromthecrypt I think I need your inputs  here

Reply:
Sorry, forgot to link the issue :) We do have some input from @codefromthecrypt in #552 though happy to change as needed.

Reply:
oh ic!

Reply:
By the way, just as a reference, in AS the default for host declarations uses the file path as the module

https://www.assemblyscript.org/concepts.html#module-imports

So it may not be as idiomatic to use env as a catch-all host module for AS programs. Still does seem a bit concerning though

Reply:
thanks for the inputs!